### PR TITLE
feat(ibc): clean up channel closure process

### PR DESCRIPTION
**Pull request by michaelfig**
refs: #9857 

## Description

Add a `finalizer` Exo to the network implementation.  The finalizer contains logic to send messages expected upon disposing of a Connection or InboundAttempt.

### Security Considerations

More systematic closing of connection objects.

### Scaling Considerations

n/a

### Documentation Considerations

Should be backwards compatible.

### Testing Considerations

PR should include an upgrade test so we know it's safe to get into master

### Upgrade Considerations

It is necessary to upgrade the network and IBC vats if they have already been deployed (as they were in `agoric-upgrade-16`).

### CODE REVIEW

File: packages/network/src/network.js [hunk 1]
```diff
@@ -10,15 +10,28 @@ import { Shape } from './shapes.js';
...
 /// <reference path="./types.js" />
 /**
- * @import {AttemptDescription, Bytes, Closable, CloseReason, Connection, ConnectionHandler, Endpoint, ListenHandler, Port, Protocol, ProtocolHandler, ProtocolImpl} from './types.js';
+ * @import {
+ *  AttemptDescription, Bytes, CloseReason, Closable,
+ *  Connection, ConnectionHandler, Endpoint, ListenHandler,
+ *  Port, Protocol, ProtocolHandler, ProtocolImpl
+ * } from './types.js';
+ * @import { PromiseVow, Remote, VowTools } from '@agoric/vow';
  */
```
Comment (rabi-siddique) [inline packages/network/src/network.js:15]: Do you think that the `*` at the beginning of each line within the JSDoc comment block might be causing the issue? When I removed these asterisks, the `network.d.ts` file generated without any problems related to the types.

After removing the asterisks:
```js
/**
@import {
 AttemptDescription, Bytes, CloseReason, Closable,
 Connection, ConnectionHandler, Endpoint, ListenHandler,
 Port, Protocol, ProtocolHandler, ProtocolImpl
} from './types.js';
@import { PromiseVow, Remote, VowTools } from '@agoric/vow';
*/

```

Comment (rabi-siddique) [reply packages/network/src/network.js:15]: <strike>

After removing the asterisks, the error message now indicates issues with non-existent module imports:
```bash
test/devices.js(1,27): error TS2307: Cannot find module '../bundles/bundle-centralSupply.js' or its corresponding type declarations.
test/devices.js(2,24): error TS2307: Cannot find module '../bundles/bundle-mintHolder.js' or its corresponding type declarations.
test/devices.js(3,27): error TS2307: Cannot find module '../bundles/bundle-provisionPool.js' or its corresponding type declarations.

```

These errors are occurring in the `packages/vats/test/devices.js` file. Trying to understand these. 

</strike>

`yarn lerna run --reject-cycles --concurrency 1 prepack` executed successfully. The error was on my end - I was working on a different machine where I hadn’t yet run `yarn build` in the `agoric-sdk` repo. Thanks to @turadg for pointing that out. :)

**Comment by 0xpatrickdev** [comment]
Restored a previous revision in [623bb31](https://github.com/Agoric/agoric-sdk/pull/9868/commits/623bb310eac9b12d86b9fae19cdc236925f5e3ec) in order for tests in #9857 to pass. I would like to understand the `initCloser` and `'closer' in disposeInfo` paths a bit more, as I would expect them to be reached in the 9857 scenarios (but they seemingly aren't).

**Comment by michaelfig** [comment]
> Restored a previous revision in [623bb31](https://github.com/Agoric/agoric-sdk/pull/9868/commits/623bb310eac9b12d86b9fae19cdc236925f5e3ec) in order for tests in #9857 to pass. I would like to understand the `initCloser` and `'closer' in disposeInfo` paths a bit more, as I would expect them to be reached in the 9857 scenarios (but they seemingly aren't).

I was trying to get the crossing closes working (where one or both sides initiate a `conn.close()`), but it didn't work well. My solution now is to always use the finalizer and keep the other state in `ibc.js` to decide which end submits the `initChannelClose`.

But as for your question, `initCloser` is only used for pending inbound attempts, which aren't likely to be triggered often by your tests.

**Review by 0xpatrickdev — APPROVED** [review]

**Pull request merged**